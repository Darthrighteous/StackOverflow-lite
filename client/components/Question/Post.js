import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';
import moment from 'moment';
import getUserDetails from '../../utils/getUserDetails';

/**
 * Post component
 */
class Post extends Component {
  constructor(props) {
    super(props);

    this.state = {
      isAddingComment: false
    };

    this.toggleCommentInput = this.toggleCommentInput.bind(this);
    this.handleUpvoteClick = this.handleUpvoteClick.bind(this);
    this.handleDownvoteClick = this.handleDownvoteClick.bind(this);
    this.handleDeleteClick = this.handleDeleteClick.bind(this);
  }

  toggleCommentInput() {
    this.setState(prevState => ({
      isAddingComment: !prevState.isAddingComment,
    }));
  }

  handleUpvoteClick(e) {
    const { type, onModifyPost, post } = this.props;
    const undo = e.target.classList.contains('active');
    const answerId = type === 'answers' ? post.id : null;
    onModifyPost('upvote', type, answerId, null, undo);
  }

  handleDownvoteClick(e) {
    const { type, onModifyPost, post } = this.props;
    const undo = e.target.classList.contains('active');
    const answerId = type === 'answers' ? post.id : null;
    onModifyPost('downvote', type, answerId, null, undo);
  }

  handleDeleteClick() {
    const { type, onDeletePost, post } = this.props;
    const answerId = type === 'answers' ? post.id : null;
    onDeletePost(type, answerId);
  }

  render() {
    const {
      type,
      isQuestionOwner,
      isLoggedIn,
      post,
      comments,
      onCommentPost,
      onModifyPost,
    } = this.props;
    const { isAddingComment } = this.state;
    const user = getUserDetails();
    const username = user ? user.username : '';
    const isOwner = username === post.username ? true : false;

    let isUpvoted = false, isDownvoted = false;
    if (user) {
      isUpvoted = type === 'answers' ?
        user.upvoted_answers.indexOf(post.id) > -1
        : user.upvoted_questions.indexOf(post.id) > -1;

      isDownvoted = type === 'answers' ?
        user.downvoted_answers.indexOf(post.id) > -1
        : user.downvoted_questions.indexOf(post.id) > -1;
    }

    return (
      <div className="post-grid">
        <div className="vote-cell">
          <button
            type="button"
            className="btn-action btn-upvote"
            onClick={isUpvoted ?
              this.handleDownvoteClick : this.handleUpvoteClick}
          >
            {(isUpvoted) ? (
              <i className="fas fa-chevron-circle-up active" />
            ) : (
              <i className="fas fa-chevron-circle-up" />
            )}
          </button>

          <span className="score-count">
            {post.score}
          </span>

          <button
            type="button"
            className="btn-action btn-downvote"
            onClick={isDownvoted ?
              this.handleUpvoteClick : this.handleDownvoteClick}
          >
            {(isDownvoted) ? (
              <i className="fas fa-chevron-circle-down active" />
            ) : (
              <i className="fas fa-chevron-circle-down" />
            )}
          </button>
        </div>
  
        <div className="post-cell">
          <div className="post-body">
            {post.body}
          </div>
          <div className="post-details">
            {(isOwner) ? (
              <div className="post-options">
                <button type="button">edit</button>
                <button
                  type="button"
                  className="btn-delete"
                  onClick={this.handleDeleteClick}
                >
                  delete
                </button>
              </div>
            ) : (<div />)}
            
            <div className="post-author-details">
              <span>{moment(post.created_at).fromNow()}</span>
              by
              <Link to={`/users/${post.username}`}>{post.username}</Link>
            </div>
          </div>
        </div>
  
        <div className="accept-cell">
          {(type === 'answers' && isQuestionOwner || post.accepted) ? (
            <button
              type="button"
              className="btn-action btn-accept"
              onClick={onModifyPost.bind(this, 'accept', type, post.id, null, null)}
            >
              {(post.accepted) ? (
                <i className="fas fa-check active" />
              ) : (
                <i className="fas fa-check" />
              )}
            </button>
          ) : (
            <div />
          )}
        </div>
  
        <div className="comment-cell">
          {(comments.length < 1) ? <div />
            : (
              <ul className="comment-list">
                {comments.map(comment => (
                  <li className="comment-item" key={comment.id}>
                    {comment.body}
                    <span>
                      &nbsp;â€” by&nbsp;
                      <Link to={`/users/${comment.username}`}>{comment.username}</Link>
                      &nbsp;
                      {moment(comment.created_at).fromNow()}
                    </span>
                  </li>
                ))
                }
              </ul>
              )
            }
         
          <div className="add-comment">
            {(!isLoggedIn) ? (
              <Link to="/signup">Sign up to add a comment</Link>
            ) : (!isAddingComment) ? (
              <button className="btn-new-comment" onClick={this.toggleCommentInput} type="button">add a comment..</button>
            ) : (
              <form className="new-comment-input" onSubmit={onCommentPost.bind(this, type, post.id)}>
                <input type="text" name="comment" placeholder="enter comment here.." />
                <button type="button" className="btn-cancel" onClick={this.toggleCommentInput}>cancel</button>
                <button type="submit" className="btn-post">post</button>
              </form>
            )}
          </div>
        </div>
      </div>
  
    );
  } 
}

Post.propTypes = {
  type: PropTypes.string.isRequired,
  isQuestionOwner: PropTypes.bool.isRequired,
  isLoggedIn: PropTypes.bool.isRequired,
  post: PropTypes.object.isRequired,
  comments: PropTypes.array.isRequired,
  onCommentPost: PropTypes.func.isRequired,
  onModifyPost: PropTypes.func.isRequired,
  onDeletePost: PropTypes.func.isRequired,
};

export default Post;
