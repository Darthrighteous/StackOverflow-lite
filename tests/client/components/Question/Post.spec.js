import React from 'react';
import { shallow } from 'enzyme';

import Post from '../../../../client/components/Question/Post';
import { mockSingleQuestionResponse } from '../../../support/mockData';

/**
 * setup for test
 * @returns {object} containing props and enzyme wrappers
 */
function setup() {
  const props = {
    type: 'questions',
    isQuestionOwner: true,
    isLoggedIn: true,
    post: mockSingleQuestionResponse.question,
    comments: mockSingleQuestionResponse.comments,
    onCommentPost: jest.fn(),
    onModifyPost: jest.fn(),
    onDeletePost: jest.fn(),
  };
  const loggedOutProps = {
    ...props, isLoggedIn: false,
  };
  const typeAnswersProps = {
    ...props,
    type: 'answers',
    comments: [],
    post: mockSingleQuestionResponse.answers[0],
  };
  const altQuestionProps = {
    ...props,
    type: 'questions',
    comments: [],
    post: mockSingleQuestionResponse.question,
  };

  const acceptedAnswerProps = {
    ...typeAnswersProps,
    post: mockSingleQuestionResponse.answers[1],
  };

  const user = {
    answer_count: 5,
    comment_count: 8,
    downvoted_answers: [1, 6, 7, 7],
    downvoted_questions: [1, 2, 4],
    email: "ja.ogunniyi@test.com",
    firstname: "Jalil",
    id: 22,
    joined_at: "2018-12-26T08:24:09.197Z",
    lastname: "Ogunniyi",
    question_count: 1,
    upvoted_answers: [1],
    upvoted_questions: [2, 1, 4],
    username: "Darthrighteous"
  };
  localStorage.setItem('user', JSON.stringify(user));
  const questionPostWrapper = shallow(<Post {...props} />);
  const altQuestionPostWrapper = shallow(<Post {...altQuestionProps} />);
  const loggedOutPostWrapper = shallow(<Post {...loggedOutProps} />);
  const answerPostWrapper = shallow(<Post {...typeAnswersProps} />);
  const acceptedAnswerPostWrapper = shallow(<Post {...acceptedAnswerProps} />);
  localStorage.clear();
  const noUserPostWrapper =shallow(<Post {...props} />);

  // const loggedOutWrapper =shallow(<Post {...loggedOutProps} />);
  // const withAnswersWrapper =shallow(<Post {...withAnswersProps} />);

  return {
    questionPostWrapper,
    altQuestionPostWrapper,
    loggedOutPostWrapper,
    answerPostWrapper,
    acceptedAnswerPostWrapper,
    noUserPostWrapper,
  };
}

describe('Post component test', () => {
  const {
    questionPostWrapper,
    altQuestionPostWrapper,
    loggedOutPostWrapper,
    answerPostWrapper,
    acceptedAnswerPostWrapper,
    noUserPostWrapper
  } = setup();

  test('snapshot test', () => {
    expect(questionPostWrapper).toMatchSnapshot();
    expect(loggedOutPostWrapper).toMatchSnapshot();
    expect(answerPostWrapper).toMatchSnapshot();
    expect(acceptedAnswerPostWrapper).toMatchSnapshot();
    expect(noUserPostWrapper).toMatchSnapshot();
  });

  test('upvote click handler', () => {
    const event = {
      target: {
        classList: {
          contains: () => (true),
        }
      }
    };
    altQuestionPostWrapper.find('.btn-upvote').simulate('click', event);
    answerPostWrapper.find('.btn-upvote').simulate('click', event);
  });

  test('downvote click handler', () => {
    const event = {
      target: {
        classList: {
          contains: () => (true),
        }
      }
    };
    altQuestionPostWrapper.find('.btn-downvote').simulate('click', event);
    answerPostWrapper.find('.btn-downvote').simulate('click', event);
  });

  test('comment input click handler', () => {
    questionPostWrapper.find('.btn-new-comment').simulate('click');
    expect(questionPostWrapper.state().isAddingComment).toEqual(true);
  });

  test('delete click handler', () => {
    altQuestionPostWrapper.find('.btn-delete').simulate('click');
    answerPostWrapper.find('.btn-delete').simulate('click');

    expect(questionPostWrapper.state().isAddingComment).toEqual(true);
  });
});
