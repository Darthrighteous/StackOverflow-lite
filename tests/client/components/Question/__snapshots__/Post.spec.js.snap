// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Post component test snapshot test 1`] = `
<div
  className="post-grid"
>
  <div
    className="vote-cell"
  >
    <button
      className="btn-action btn-upvote"
      onClick={[Function]}
      type="button"
    >
      <i
        className="fas fa-chevron-circle-up active"
      />
    </button>
    <span
      className="score-count"
    >
      0
    </span>
    <button
      className="btn-action btn-downvote"
      onClick={[Function]}
      type="button"
    >
      <i
        className="fas fa-chevron-circle-down active"
      />
    </button>
  </div>
  <div
    className="post-cell"
  >
    <div
      className="post-body"
    >
      What is the point of celebrating when we win? Is it cause we feel happy? is it cause we don't know how to act when we feel happy. I am coming from a place of winning all the time and I'm not sure what to do about this, I feel like I 'NEED' to celebrate. This doesn't feel normal to me. I don't want to celebrate because I want to, it feels more like society has programmed me to NEED to celebrate. Why can't I just go about my day normally?
    </div>
    <div
      className="post-details"
    >
      <div
        className="post-options"
      >
        <button
          type="button"
        >
          edit
        </button>
        <button
          className="btn-delete"
          onClick={[Function]}
          type="button"
        >
          delete
        </button>
      </div>
      <div
        className="post-author-details"
      >
        <span>
          7 days ago
        </span>
        by
        <Link
          replace={false}
          to="/users/Darthrighteous"
        >
          Darthrighteous
        </Link>
      </div>
    </div>
  </div>
  <div
    className="accept-cell"
  >
    <div />
  </div>
  <div
    className="comment-cell"
  >
    <ul
      className="comment-list"
    >
      <li
        className="comment-item"
        key="3"
      >
        this is a very serious question, please answer if you know
        <span>
           — by 
          <Link
            replace={false}
            to="/users/Darthrighteous"
          >
            Darthrighteous
          </Link>
           
          5 days ago
        </span>
      </li>
      <li
        className="comment-item"
        key="4"
      >
        This is a new comment that I am posting for the sake of it
        <span>
           — by 
          <Link
            replace={false}
            to="/users/Darthrighteous"
          >
            Darthrighteous
          </Link>
           
          5 days ago
        </span>
      </li>
    </ul>
    <div
      className="add-comment"
    >
      <button
        className="btn-new-comment"
        onClick={[Function]}
        type="button"
      >
        add a comment..
      </button>
    </div>
  </div>
</div>
`;

exports[`Post component test snapshot test 2`] = `
<div
  className="post-grid"
>
  <div
    className="vote-cell"
  >
    <button
      className="btn-action btn-upvote"
      onClick={[Function]}
      type="button"
    >
      <i
        className="fas fa-chevron-circle-up active"
      />
    </button>
    <span
      className="score-count"
    >
      0
    </span>
    <button
      className="btn-action btn-downvote"
      onClick={[Function]}
      type="button"
    >
      <i
        className="fas fa-chevron-circle-down active"
      />
    </button>
  </div>
  <div
    className="post-cell"
  >
    <div
      className="post-body"
    >
      What is the point of celebrating when we win? Is it cause we feel happy? is it cause we don't know how to act when we feel happy. I am coming from a place of winning all the time and I'm not sure what to do about this, I feel like I 'NEED' to celebrate. This doesn't feel normal to me. I don't want to celebrate because I want to, it feels more like society has programmed me to NEED to celebrate. Why can't I just go about my day normally?
    </div>
    <div
      className="post-details"
    >
      <div
        className="post-options"
      >
        <button
          type="button"
        >
          edit
        </button>
        <button
          className="btn-delete"
          onClick={[Function]}
          type="button"
        >
          delete
        </button>
      </div>
      <div
        className="post-author-details"
      >
        <span>
          7 days ago
        </span>
        by
        <Link
          replace={false}
          to="/users/Darthrighteous"
        >
          Darthrighteous
        </Link>
      </div>
    </div>
  </div>
  <div
    className="accept-cell"
  >
    <div />
  </div>
  <div
    className="comment-cell"
  >
    <ul
      className="comment-list"
    >
      <li
        className="comment-item"
        key="3"
      >
        this is a very serious question, please answer if you know
        <span>
           — by 
          <Link
            replace={false}
            to="/users/Darthrighteous"
          >
            Darthrighteous
          </Link>
           
          5 days ago
        </span>
      </li>
      <li
        className="comment-item"
        key="4"
      >
        This is a new comment that I am posting for the sake of it
        <span>
           — by 
          <Link
            replace={false}
            to="/users/Darthrighteous"
          >
            Darthrighteous
          </Link>
           
          5 days ago
        </span>
      </li>
    </ul>
    <div
      className="add-comment"
    >
      <Link
        replace={false}
        to="/signup"
      >
        Sign up to add a comment
      </Link>
    </div>
  </div>
</div>
`;

exports[`Post component test snapshot test 3`] = `
<div
  className="post-grid"
>
  <div
    className="vote-cell"
  >
    <button
      className="btn-action btn-upvote"
      onClick={[Function]}
      type="button"
    >
      <i
        className="fas fa-chevron-circle-up"
      />
    </button>
    <span
      className="score-count"
    >
      0
    </span>
    <button
      className="btn-action btn-downvote"
      onClick={[Function]}
      type="button"
    >
      <i
        className="fas fa-chevron-circle-down"
      />
    </button>
  </div>
  <div
    className="post-cell"
  >
    <div
      className="post-body"
    >
      Well, we celebrate winning cause it makes us feel good. You shouldn't worry about why we do it. All that matters is your happiness whether you've been programmed to do it or not.
    </div>
    <div
      className="post-details"
    >
      <div
        className="post-options"
      >
        <button
          type="button"
        >
          edit
        </button>
        <button
          className="btn-delete"
          onClick={[Function]}
          type="button"
        >
          delete
        </button>
      </div>
      <div
        className="post-author-details"
      >
        <span>
          4 days ago
        </span>
        by
        <Link
          replace={false}
          to="/users/Darthrighteous"
        >
          Darthrighteous
        </Link>
      </div>
    </div>
  </div>
  <div
    className="accept-cell"
  >
    <button
      className="btn-action btn-accept"
      onClick={[Function]}
      type="button"
    >
      <i
        className="fas fa-check"
      />
    </button>
  </div>
  <div
    className="comment-cell"
  >
    <div />
    <div
      className="add-comment"
    >
      <button
        className="btn-new-comment"
        onClick={[Function]}
        type="button"
      >
        add a comment..
      </button>
    </div>
  </div>
</div>
`;

exports[`Post component test snapshot test 4`] = `
<div
  className="post-grid"
>
  <div
    className="vote-cell"
  >
    <button
      className="btn-action btn-upvote"
      onClick={[Function]}
      type="button"
    >
      <i
        className="fas fa-chevron-circle-up"
      />
    </button>
    <span
      className="score-count"
    >
      0
    </span>
    <button
      className="btn-action btn-downvote"
      onClick={[Function]}
      type="button"
    >
      <i
        className="fas fa-chevron-circle-down active"
      />
    </button>
  </div>
  <div
    className="post-cell"
  >
    <div
      className="post-body"
    >
      Duplicated. Well, we celebrate winning cause it makes us feel good. You shouldn't worry about why we do it. All that matters is your happiness whether you've been programmed to do it or not.
    </div>
    <div
      className="post-details"
    >
      <div
        className="post-options"
      >
        <button
          type="button"
        >
          edit
        </button>
        <button
          className="btn-delete"
          onClick={[Function]}
          type="button"
        >
          delete
        </button>
      </div>
      <div
        className="post-author-details"
      >
        <span>
          4 days ago
        </span>
        by
        <Link
          replace={false}
          to="/users/Darthrighteous"
        >
          Darthrighteous
        </Link>
      </div>
    </div>
  </div>
  <div
    className="accept-cell"
  >
    <button
      className="btn-action btn-accept"
      onClick={[Function]}
      type="button"
    >
      <i
        className="fas fa-check active"
      />
    </button>
  </div>
  <div
    className="comment-cell"
  >
    <div />
    <div
      className="add-comment"
    >
      <button
        className="btn-new-comment"
        onClick={[Function]}
        type="button"
      >
        add a comment..
      </button>
    </div>
  </div>
</div>
`;

exports[`Post component test snapshot test 5`] = `
<div
  className="post-grid"
>
  <div
    className="vote-cell"
  >
    <button
      className="btn-action btn-upvote"
      onClick={[Function]}
      type="button"
    >
      <i
        className="fas fa-chevron-circle-up"
      />
    </button>
    <span
      className="score-count"
    >
      0
    </span>
    <button
      className="btn-action btn-downvote"
      onClick={[Function]}
      type="button"
    >
      <i
        className="fas fa-chevron-circle-down"
      />
    </button>
  </div>
  <div
    className="post-cell"
  >
    <div
      className="post-body"
    >
      What is the point of celebrating when we win? Is it cause we feel happy? is it cause we don't know how to act when we feel happy. I am coming from a place of winning all the time and I'm not sure what to do about this, I feel like I 'NEED' to celebrate. This doesn't feel normal to me. I don't want to celebrate because I want to, it feels more like society has programmed me to NEED to celebrate. Why can't I just go about my day normally?
    </div>
    <div
      className="post-details"
    >
      <div />
      <div
        className="post-author-details"
      >
        <span>
          7 days ago
        </span>
        by
        <Link
          replace={false}
          to="/users/Darthrighteous"
        >
          Darthrighteous
        </Link>
      </div>
    </div>
  </div>
  <div
    className="accept-cell"
  >
    <div />
  </div>
  <div
    className="comment-cell"
  >
    <ul
      className="comment-list"
    >
      <li
        className="comment-item"
        key="3"
      >
        this is a very serious question, please answer if you know
        <span>
           — by 
          <Link
            replace={false}
            to="/users/Darthrighteous"
          >
            Darthrighteous
          </Link>
           
          5 days ago
        </span>
      </li>
      <li
        className="comment-item"
        key="4"
      >
        This is a new comment that I am posting for the sake of it
        <span>
           — by 
          <Link
            replace={false}
            to="/users/Darthrighteous"
          >
            Darthrighteous
          </Link>
           
          5 days ago
        </span>
      </li>
    </ul>
    <div
      className="add-comment"
    >
      <button
        className="btn-new-comment"
        onClick={[Function]}
        type="button"
      >
        add a comment..
      </button>
    </div>
  </div>
</div>
`;
